File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/CreateTrustedCircuit.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * CreateTrustedCircuit
 * Sent to establish a trust relationship between two components.
 * Only sent in response to a DenyTrustedCircuit message.
 *
 * <p>Template: {@code CreateTrustedCircuit Low 392 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_create_trusted_circuit()} in indra/llmessage/message.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class CreateTrustedCircuit extends SLMessage {
    public DataBlock DataBlock_Field;

    /** Block DataBlock, Single. */
    public static class DataBlock {
        public byte[] Digest; // Fixed 32 - 32 hex digits == 1 MD5 Digest
        public UUID EndPointID; // LLUUID
    }

    public CreateTrustedCircuit() {
        this.zeroCoded = false;
        this.DataBlock_Field = new DataBlock();
    }

    @Override
    public int CalcPayloadSize() {
        return 52;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleCreateTrustedCircuit(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 392 (CreateTrustedCircuit).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x01);
        byteBuffer.put((byte) 0x88);
        packUUID(byteBuffer, this.DataBlock_Field.EndPointID);
        packFixed(byteBuffer, this.DataBlock_Field.Digest, 32);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.DataBlock_Field.EndPointID = unpackUUID(byteBuffer);
        this.DataBlock_Field.Digest = unpackFixed(byteBuffer, 32);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/CreateTrustedCircuit.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * CreateTrustedCircuit
 * Sent to establish a trust relationship between two components.
 * Only sent in response to a DenyTrustedCircuit message.
 *
 * <p>Template: {@code CreateTrustedCircuit Low 392 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_create_trusted_circuit()} in indra/llmessage/message.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class CreateTrustedCircuit : SLMessage() {
    @JvmField var DataBlock_Field: DataBlock = DataBlock()

    /** Block DataBlock, Single. */
    open class DataBlock {
        @JvmField var Digest: ByteArray? = null
        @JvmField var EndPointID: UUID? = null
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return 52
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleCreateTrustedCircuit(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 392 (CreateTrustedCircuit).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x01).toByte())
        byteBuffer.put((0x88).toByte())
        packUUID(byteBuffer, DataBlock_Field.EndPointID)
        packFixed(byteBuffer, DataBlock_Field.Digest, 32)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        DataBlock_Field.EndPointID = unpackUUID(byteBuffer)
        DataBlock_Field.Digest = unpackFixed(byteBuffer, 32)
    }
}
